R/el_cuml.R

```diff
@@ -0,0 +1,132 @@
+#'                from
+#'
+#' @return a cumulative edgelist in \code{data.frame} form with 4 columns:
+#'          head and tail, the unique_ids (see \code{get_unique_ids}) of the
+#'          nodes on the edge. Start and stop, the timestep where the edges
+#'          started and stopped.
+#'
+#' @export
+get_cumulative_edgelist <- function(dat, network) {
+  el_cuml <- dat[["el_cuml"]][[network]]
+
+  if(is.null(el_cuml)) {
+    el_cuml <- data.frame(
+      head  = numeric(0),
+      tail  = numeric(0),
+      start = numeric(0),
+      stop  = numeric(0)
+    )
+  }
+
+  return(el_cuml)
+}
+
+#' @title Update a Cumulative Edgelist of the Specified Network
+#'
+#' @param dat a Master list object of network models
+#' @param at the current timestep
+#' @param network the index of the network whose cumulative edgelist will be
+#'                updated
+#'
+#' @section Truncation:
+#' To avoid storing a cumulative edgelist to long, the "truncate.el_cuml"
+#' control value defines a number of steps after which an edge that is no longer
+#' active is truncated out of the cummulative edgelist.
+#'
+#' @return an updated Master list object of network models
+#'
+#' @export
+update_cumulative_edgelist <- function(dat, at, network) {
+  truncate.el_cuml <- get_control(
+    dat, "truncate.el_cuml", override.null.error = TRUE)
+  truncate.el_cuml <- if (is.null(truncate.el_cuml)) Inf else truncate.el_cuml
+
+  el <- get_edgelist(dat, at, network)
+  el_cuml <- get_cumulative_edgelist(dat, network)
+
+  el <- data.frame(
+    head = get_unique_ids(dat, el[, 1]),
+    tail = get_unique_ids(dat, el[, 2]),
+    current = TRUE
+  )
+
+  el_cuml <- merge(el_cuml, el, by = c("head", "tail"), all = TRUE)
+
+  new_edges <- is.na(el_cuml[["start"]])
+  if (any(new_edges)) {
+    el_cuml[new_edges, ][["start"]] <- at
+  }
+
+  terminated_edges <- is.na(el_cuml[["current"]]) & is.na(el_cuml[["stop"]])
+  if (any(terminated_edges)) {
+    el_cuml[terminated_edges, ][["stop"]] <- at - 1
+  }
+
+  if (truncate.el_cuml != Inf) {
+    rel.age <- at - el_cuml[["stop"]]
+    rel.age <- ifelse(is.na(rel.age), 0, rel.age)
+    el_cuml <- el_cuml[rel.age <= truncate.el_cuml, ]
+  }
+
+  dat[["el_cuml"]][[network]] <- el_cuml[, c("head", "tail", "start", "stop")]
+
+  return(dat)
+}
+
+#' @title Get the Cumulative Edgelists of a Model in \code{data.frame} format
+#'
+#' @param dat a Master list object of network models
+#' @param networks the indexes of the networks to extract the cumulative
+#'                 edgelists from. If NULL (default) extract all the cumulative
+#'                 edgelists.
+#'
+#' @return a \code{data.frame} with 5 columns:
+#'          head and tail, the unique_ids (see \code{get_unique_ids}) of the
+#'          nodes on the edge. Start and stop, the timestep where the edges
+#'          started and stopped. network, the network on which the edge is.
+#'
+#' @export
+get_cumulative_edgelists_df <- function(dat, networks = NULL) {
```

Comment:
is this just for post processing of the netsim object? Just curious where this is used.

Reply:
Right now it is not used. But I was thinking of what we do in CombPrevNet where this is the format we would want to work with. In general I feel that having things stored individually inside a `list` gives the best flexibility but combined `data.frame`s are better suited for analysis and filtering for custom module. As it can be tricky to consistently combine `data.frame`s into one without errors (especially in corner cases, where one is empty for instance) I thought it would be useful to have the helper function directly in so we can test it thoroughly and module builders will be able to use it without risking unforeseen errors.